{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DataTransformerRegistry.enable('default')"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd \n",
    "import altair as alt\n",
    "import re \n",
    "import ntpath \n",
    "from commons import data_processing\n",
    "from commons.DataProcessors import msfragger\n",
    "from commons.APICallers import uniprot\n",
    "\n",
    "alt.data_transformers.disable_max_rows()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "MSF Processor constructed on the following data:\n",
       "./data/fractions/PGC\\F1\\psm.tsv\n",
       "./data/fractions/PGC\\F2\\psm.tsv\n",
       "./data/fractions/PGC\\F3\\psm.tsv\n",
       "./data/fractions/PGC\\F4\\psm.tsv\n",
       "./data/fractions/PGC\\F5\\psm.tsv\n",
       "./data/fractions/PGC\\F6\\psm.tsv\n",
       "./data/fractions/PGC\\F7\\psm.tsv\n",
       "./data/fractions/PGC\\F8\\psm.tsv\n",
       "./data/fractions/RPLC\\F1\\psm.tsv\n",
       "./data/fractions/RPLC\\F2\\psm.tsv\n",
       "./data/fractions/RPLC\\F3\\psm.tsv\n",
       "./data/fractions/RPLC\\F4\\psm.tsv\n",
       "./data/fractions/RPLC\\F5\\psm.tsv\n",
       "./data/fractions/RPLC\\F6\\psm.tsv\n",
       "./data/fractions/RPLC\\F7\\psm.tsv\n",
       "./data/fractions/RPLC\\F8\\psm.tsv"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "files = data_processing.get_files(r'./data/fractions/', exts=['psm.tsv'])\n",
    "\n",
    "# use msfragger module to process data\n",
    "msf = None\n",
    "for file in files:\n",
    "    m = msfragger.msf_processor([file])\n",
    "\n",
    "#     # extract fraction and column info\n",
    "    dirname = ntpath.dirname(file)\n",
    "    spl = re.split(r'[\\\\\\/]', dirname)\n",
    "    column, fraction = spl[-2:]\n",
    "    m.add_special_column('fraction', fraction)\n",
    "    m.add_special_column('column', column)\n",
    "\n",
    "    if msf is None:\n",
    "        msf = m\n",
    "    else:\n",
    "        msf.join_processors(m)\n",
    "\n",
    "# grab data\n",
    "df = msf.data\n",
    "msf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>spectrum</th>\n",
       "      <th>spectrum_file</th>\n",
       "      <th>peptide</th>\n",
       "      <th>modified_peptide</th>\n",
       "      <th>prev_aa</th>\n",
       "      <th>next_aa</th>\n",
       "      <th>peptide_length</th>\n",
       "      <th>charge</th>\n",
       "      <th>retention</th>\n",
       "      <th>observed_mass</th>\n",
       "      <th>...</th>\n",
       "      <th>is_unique</th>\n",
       "      <th>protein</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>entry_name</th>\n",
       "      <th>gene</th>\n",
       "      <th>protein_description</th>\n",
       "      <th>mapped_genes</th>\n",
       "      <th>mapped_proteins</th>\n",
       "      <th>fraction</th>\n",
       "      <th>column</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>20221109_GD_PGC_F1_run1.02562.02562.2</td>\n",
       "      <td>D:\\MSFragger\\20221109\\MSFragger_PGC_Frations\\F...</td>\n",
       "      <td>ATGAATPKK</td>\n",
       "      <td>ATGAATPKK</td>\n",
       "      <td>K</td>\n",
       "      <td>S</td>\n",
       "      <td>9</td>\n",
       "      <td>2</td>\n",
       "      <td>758.9805</td>\n",
       "      <td>843.4767</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>sp|P10412|H14_HUMAN</td>\n",
       "      <td>P10412</td>\n",
       "      <td>H14_HUMAN</td>\n",
       "      <td>HIST1H1E</td>\n",
       "      <td>Histone H1.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F1</td>\n",
       "      <td>PGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>20221109_GD_PGC_F1_run1.02591.02591.2</td>\n",
       "      <td>D:\\MSFragger\\20221109\\MSFragger_PGC_Frations\\F...</td>\n",
       "      <td>ATTKPPPAK</td>\n",
       "      <td>ATTKPPPAK</td>\n",
       "      <td>K</td>\n",
       "      <td>K</td>\n",
       "      <td>9</td>\n",
       "      <td>2</td>\n",
       "      <td>768.1130</td>\n",
       "      <td>909.5245</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>sp|Q14978|NOLC1_HUMAN</td>\n",
       "      <td>Q14978</td>\n",
       "      <td>NOLC1_HUMAN</td>\n",
       "      <td>NOLC1</td>\n",
       "      <td>Nucleolar and coiled-body phosphoprotein 1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F1</td>\n",
       "      <td>PGC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 39 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                spectrum  \\\n",
       "0  20221109_GD_PGC_F1_run1.02562.02562.2   \n",
       "1  20221109_GD_PGC_F1_run1.02591.02591.2   \n",
       "\n",
       "                                       spectrum_file    peptide  \\\n",
       "0  D:\\MSFragger\\20221109\\MSFragger_PGC_Frations\\F...  ATGAATPKK   \n",
       "1  D:\\MSFragger\\20221109\\MSFragger_PGC_Frations\\F...  ATTKPPPAK   \n",
       "\n",
       "  modified_peptide prev_aa next_aa  peptide_length  charge  retention  \\\n",
       "0        ATGAATPKK       K       S               9       2   758.9805   \n",
       "1        ATTKPPPAK       K       K               9       2   768.1130   \n",
       "\n",
       "   observed_mass  ...  is_unique                protein  protein_id  \\\n",
       "0       843.4767  ...       True    sp|P10412|H14_HUMAN      P10412   \n",
       "1       909.5245  ...       True  sp|Q14978|NOLC1_HUMAN      Q14978   \n",
       "\n",
       "    entry_name      gene                         protein_description  \\\n",
       "0    H14_HUMAN  HIST1H1E                                Histone H1.4   \n",
       "1  NOLC1_HUMAN     NOLC1  Nucleolar and coiled-body phosphoprotein 1   \n",
       "\n",
       "   mapped_genes  mapped_proteins  fraction  column  \n",
       "0           NaN              NaN        F1     PGC  \n",
       "1           NaN              NaN        F1     PGC  \n",
       "\n",
       "[2 rows x 39 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "no_mods = df[df.modified_peptide.isna()].index\n",
    "df.loc[no_mods, \"modified_peptide\"] = df.loc[no_mods, \"peptide\"]\n",
    "df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "avg = df.groupby(['column', 'fraction', 'modified_peptide', 'peptide',]).mean()\n",
    "avg.loc[:, 'retention_min'] = avg.retention/60"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "column_colors = alt.Color('column:N',\n",
    "    scale=alt.Scale(\n",
    "        domain=[\"PGC\", \"RPLC\"],\n",
    "        range=[\"#D56E3B\", \"#58728C\"]\n",
    "    ))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "overlap = 1\n",
    "\n",
    "alt.Chart(avg.reset_index(), height=40).mark_area(\n",
    "    opacity=1,\n",
    "    stroke='lightgray',\n",
    "    strokeWidth=0.5\n",
    ").transform_density(\n",
    "    'retention_min',\n",
    "    as_ = ['retention_min', 'density'],\n",
    "    groupby=['fraction', 'column',]\n",
    ").encode(\n",
    "    x=alt.X('retention_min:Q', title='Retention Time (min)',\n",
    "        axis=alt.Axis(grid=False)),\n",
    "    y=alt.Y('density:Q', axis=None,\n",
    "        scale=alt.Scale(range=[40, -5])),\n",
    "    color=column_colors\n",
    ").facet(\n",
    "    row=alt.Row('fraction:N', title=None,\n",
    "        header=alt.Header(labelAngle=0)),\n",
    "    column=alt.Column('column:N', title=None)\n",
    ").configure_facet(\n",
    "    spacing=0\n",
    ").configure_view(\n",
    "    stroke=None\n",
    ").properties(\n",
    "    bounds='flush'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = avg.reset_index()\n",
    "# d = d.drop_duplicates(['modified_peptide', 'fraction'])\n",
    "total = alt.Chart(d).mark_bar().encode(\n",
    "    x=alt.X('column:N', title='', \n",
    "        axis=alt.Axis(labels=False, ticks=False)),\n",
    "    y=alt.Y('distinct(modified_peptide)', title='Unique Peptides'),\n",
    "    color=column_colors,\n",
    "    column=alt.Column('fraction:N')\n",
    ").properties(\n",
    "    height=250,\n",
    "    width=25\n",
    ")\n",
    "\n",
    "unique_frac = alt.Chart(d.drop_duplicates(['modified_peptide', 'fraction'])).mark_bar().encode(\n",
    "    x=alt.X('column:N', title='', \n",
    "        axis=alt.Axis(labels=False, ticks=False)),\n",
    "    y=alt.Y('distinct(modified_peptide)', title='Unique Peptides'),\n",
    "    color=column_colors,\n",
    "    column=alt.Column('fraction:N')\n",
    ").properties(\n",
    "    height=250,\n",
    "    width=25\n",
    ")\n",
    "\n",
    "unique = alt.Chart(d.drop_duplicates(['modified_peptide'])).mark_bar().encode(\n",
    "    x=alt.X('column:N', title='', \n",
    "        axis=alt.Axis(labels=False, ticks=False)),\n",
    "    y=alt.Y('distinct(modified_peptide)', title='Unique Peptides'),\n",
    "    color=column_colors,\n",
    "    column=alt.Column('fraction:N')\n",
    ").properties(\n",
    "    height=250,\n",
    "    width=25\n",
    ")\n",
    "\n",
    "total | unique_frac | unique"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from modlamp.descriptors import PeptideDescriptor\n",
    "\n",
    "def pour(seq):\n",
    "    desc = PeptideDescriptor(seq, 'gravy')\n",
    "    desc.calculate_global()\n",
    "    return desc.descriptor[0][0]\n",
    "\n",
    "df.loc[:, 'gravy'] = df.peptide.map(pour)\n",
    "avg = df.groupby(['column', 'modified_peptide', 'peptide']).mean()\n",
    "avg.loc[:, 'retention_min'] = avg.retention/60"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = avg.reset_index()\n",
    "# d.drop_duplicates('modified_peptide', inplace=True)\n",
    "base = alt.Chart(d).encode(\n",
    "    x=alt.X('retention_min:Q',\n",
    "        bin=alt.Bin(\n",
    "            step=5\n",
    "        )),\n",
    "    y=alt.Y('mean(gravy):Q'),\n",
    "    color='column:N'\n",
    ")\n",
    "\n",
    "base.mark_line(interpolate='basis') + base.mark_errorband(extent='ci', interpolate='basis')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = avg.reset_index()\n",
    "d.drop_duplicates('modified_peptide', inplace=True)\n",
    "alt.Chart(d).mark_boxplot().encode(\n",
    "    x='column:N',\n",
    "    y='peptide_length:Q'\n",
    ")\n",
    "\n",
    "# import seaborn as sns \n",
    "# sns.violinplot(x='column', y='gravy', data=avg.reset_index())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5750\n"
     ]
    },
    {
     "ename": "ChunkedEncodingError",
     "evalue": "(\"Connection broken: InvalidChunkLength(got length b'', 0 bytes read)\", InvalidChunkLength(got length b'', 0 bytes read))",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "File \u001b[1;32mc:\\Users\\graha\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\response.py:700\u001b[0m, in \u001b[0;36mHTTPResponse._update_chunk_length\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    699\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m--> 700\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mchunk_left \u001b[39m=\u001b[39m \u001b[39mint\u001b[39;49m(line, \u001b[39m16\u001b[39;49m)\n\u001b[0;32m    701\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mValueError\u001b[39;00m:\n\u001b[0;32m    702\u001b[0m     \u001b[39m# Invalid chunked protocol response, abort.\u001b[39;00m\n",
      "\u001b[1;31mValueError\u001b[0m: invalid literal for int() with base 16: b''",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[1;31mInvalidChunkLength\u001b[0m                        Traceback (most recent call last)",
      "File \u001b[1;32mc:\\Users\\graha\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\response.py:441\u001b[0m, in \u001b[0;36mHTTPResponse._error_catcher\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    440\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m--> 441\u001b[0m     \u001b[39myield\u001b[39;00m\n\u001b[0;32m    443\u001b[0m \u001b[39mexcept\u001b[39;00m SocketTimeout:\n\u001b[0;32m    444\u001b[0m     \u001b[39m# FIXME: Ideally we'd like to include the url in the ReadTimeoutError but\u001b[39;00m\n\u001b[0;32m    445\u001b[0m     \u001b[39m# there is yet no clean way to get at it from this context.\u001b[39;00m\n",
      "File \u001b[1;32mc:\\Users\\graha\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\response.py:767\u001b[0m, in \u001b[0;36mHTTPResponse.read_chunked\u001b[1;34m(self, amt, decode_content)\u001b[0m\n\u001b[0;32m    766\u001b[0m \u001b[39mwhile\u001b[39;00m \u001b[39mTrue\u001b[39;00m:\n\u001b[1;32m--> 767\u001b[0m     \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_update_chunk_length()\n\u001b[0;32m    768\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mchunk_left \u001b[39m==\u001b[39m \u001b[39m0\u001b[39m:\n",
      "File \u001b[1;32mc:\\Users\\graha\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\response.py:704\u001b[0m, in \u001b[0;36mHTTPResponse._update_chunk_length\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    703\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mclose()\n\u001b[1;32m--> 704\u001b[0m \u001b[39mraise\u001b[39;00m InvalidChunkLength(\u001b[39mself\u001b[39m, line)\n",
      "\u001b[1;31mInvalidChunkLength\u001b[0m: InvalidChunkLength(got length b'', 0 bytes read)",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[1;31mProtocolError\u001b[0m                             Traceback (most recent call last)",
      "File \u001b[1;32mc:\\Users\\graha\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\requests\\models.py:816\u001b[0m, in \u001b[0;36mResponse.iter_content.<locals>.generate\u001b[1;34m()\u001b[0m\n\u001b[0;32m    815\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m--> 816\u001b[0m     \u001b[39myield from\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mraw\u001b[39m.\u001b[39mstream(chunk_size, decode_content\u001b[39m=\u001b[39m\u001b[39mTrue\u001b[39;00m)\n\u001b[0;32m    817\u001b[0m \u001b[39mexcept\u001b[39;00m ProtocolError \u001b[39mas\u001b[39;00m e:\n",
      "File \u001b[1;32mc:\\Users\\graha\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\response.py:575\u001b[0m, in \u001b[0;36mHTTPResponse.stream\u001b[1;34m(self, amt, decode_content)\u001b[0m\n\u001b[0;32m    574\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mchunked \u001b[39mand\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39msupports_chunked_reads():\n\u001b[1;32m--> 575\u001b[0m     \u001b[39mfor\u001b[39;00m line \u001b[39min\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mread_chunked(amt, decode_content\u001b[39m=\u001b[39mdecode_content):\n\u001b[0;32m    576\u001b[0m         \u001b[39myield\u001b[39;00m line\n",
      "File \u001b[1;32mc:\\Users\\graha\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\response.py:755\u001b[0m, in \u001b[0;36mHTTPResponse.read_chunked\u001b[1;34m(self, amt, decode_content)\u001b[0m\n\u001b[0;32m    750\u001b[0m     \u001b[39mraise\u001b[39;00m BodyNotHttplibCompatible(\n\u001b[0;32m    751\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39mBody should be http.client.HTTPResponse like. \u001b[39m\u001b[39m\"\u001b[39m\n\u001b[0;32m    752\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39mIt should have have an fp attribute which returns raw chunks.\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[0;32m    753\u001b[0m     )\n\u001b[1;32m--> 755\u001b[0m \u001b[39mwith\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_error_catcher():\n\u001b[0;32m    756\u001b[0m     \u001b[39m# Don't bother reading the body of a HEAD request.\u001b[39;00m\n\u001b[0;32m    757\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_original_response \u001b[39mand\u001b[39;00m is_response_to_head(\u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_original_response):\n",
      "File \u001b[1;32mc:\\Users\\graha\\AppData\\Local\\Programs\\Python\\Python310\\lib\\contextlib.py:153\u001b[0m, in \u001b[0;36m_GeneratorContextManager.__exit__\u001b[1;34m(self, typ, value, traceback)\u001b[0m\n\u001b[0;32m    152\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m--> 153\u001b[0m     \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mgen\u001b[39m.\u001b[39;49mthrow(typ, value, traceback)\n\u001b[0;32m    154\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mStopIteration\u001b[39;00m \u001b[39mas\u001b[39;00m exc:\n\u001b[0;32m    155\u001b[0m     \u001b[39m# Suppress StopIteration *unless* it's the same exception that\u001b[39;00m\n\u001b[0;32m    156\u001b[0m     \u001b[39m# was passed to throw().  This prevents a StopIteration\u001b[39;00m\n\u001b[0;32m    157\u001b[0m     \u001b[39m# raised inside the \"with\" statement from being suppressed.\u001b[39;00m\n",
      "File \u001b[1;32mc:\\Users\\graha\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\response.py:458\u001b[0m, in \u001b[0;36mHTTPResponse._error_catcher\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    456\u001b[0m \u001b[39mexcept\u001b[39;00m (HTTPException, SocketError) \u001b[39mas\u001b[39;00m e:\n\u001b[0;32m    457\u001b[0m     \u001b[39m# This includes IncompleteRead.\u001b[39;00m\n\u001b[1;32m--> 458\u001b[0m     \u001b[39mraise\u001b[39;00m ProtocolError(\u001b[39m\"\u001b[39m\u001b[39mConnection broken: \u001b[39m\u001b[39m%r\u001b[39;00m\u001b[39m\"\u001b[39m \u001b[39m%\u001b[39m e, e)\n\u001b[0;32m    460\u001b[0m \u001b[39m# If no exception is thrown, we should avoid cleaning up\u001b[39;00m\n\u001b[0;32m    461\u001b[0m \u001b[39m# unnecessarily.\u001b[39;00m\n",
      "\u001b[1;31mProtocolError\u001b[0m: (\"Connection broken: InvalidChunkLength(got length b'', 0 bytes read)\", InvalidChunkLength(got length b'', 0 bytes read))",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[1;31mChunkedEncodingError\u001b[0m                      Traceback (most recent call last)",
      "\u001b[1;32mc:\\Users\\graha\\code\\python\\pgc-rplc-frac-profiling\\2_fractions_data_workup.ipynb Cell 11\u001b[0m in \u001b[0;36m<cell line: 6>\u001b[1;34m()\u001b[0m\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Users/graha/code/python/pgc-rplc-frac-profiling/2_fractions_data_workup.ipynb#X13sZmlsZQ%3D%3D?line=5'>6</a>\u001b[0m \u001b[39mfor\u001b[39;00m i \u001b[39min\u001b[39;00m \u001b[39mrange\u001b[39m(\u001b[39m0\u001b[39m, \u001b[39mlen\u001b[39m(accs)\u001b[39m+\u001b[39m\u001b[39m1000\u001b[39m, \u001b[39m1000\u001b[39m):\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Users/graha/code/python/pgc-rplc-frac-profiling/2_fractions_data_workup.ipynb#X13sZmlsZQ%3D%3D?line=6'>7</a>\u001b[0m     sent \u001b[39m=\u001b[39m accs[i:i\u001b[39m+\u001b[39m\u001b[39m1000\u001b[39m]\n\u001b[1;32m----> <a href='vscode-notebook-cell:/c%3A/Users/graha/code/python/pgc-rplc-frac-profiling/2_fractions_data_workup.ipynb#X13sZmlsZQ%3D%3D?line=7'>8</a>\u001b[0m     call \u001b[39m=\u001b[39m uniprot\u001b[39m.\u001b[39;49msend_accessions(sent)\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Users/graha/code/python/pgc-rplc-frac-profiling/2_fractions_data_workup.ipynb#X13sZmlsZQ%3D%3D?line=8'>9</a>\u001b[0m     resp \u001b[39m=\u001b[39m uniprot\u001b[39m.\u001b[39mparse_response(call, wanted_value\u001b[39m=\u001b[39m\u001b[39m'\u001b[39m\u001b[39mgene\u001b[39m\u001b[39m'\u001b[39m)\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/graha/code/python/pgc-rplc-frac-profiling/2_fractions_data_workup.ipynb#X13sZmlsZQ%3D%3D?line=9'>10</a>\u001b[0m     loc_map\u001b[39m.\u001b[39mupdate(resp)\n",
      "File \u001b[1;32mc:\\Users\\graha\\code\\python\\pgc-rplc-frac-profiling\\commons\\APICallers\\uniprot.py:28\u001b[0m, in \u001b[0;36msend_accessions\u001b[1;34m(protein_accession)\u001b[0m\n\u001b[0;32m     25\u001b[0m base_url \u001b[39m=\u001b[39m \u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mhttps://rest.uniprot.org/uniprotkb/accessions?accessions=\u001b[39m\u001b[39m{\u001b[39;00maccs\u001b[39m}\u001b[39;00m\u001b[39m\"\u001b[39m\n\u001b[0;32m     27\u001b[0m \u001b[39m# get response\u001b[39;00m\n\u001b[1;32m---> 28\u001b[0m r \u001b[39m=\u001b[39m requests\u001b[39m.\u001b[39;49mget(base_url)\n\u001b[0;32m     30\u001b[0m \u001b[39m# dump into json object\u001b[39;00m\n\u001b[0;32m     32\u001b[0m resp \u001b[39m=\u001b[39m json\u001b[39m.\u001b[39mloads(r\u001b[39m.\u001b[39mtext)\n",
      "File \u001b[1;32mc:\\Users\\graha\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\requests\\api.py:73\u001b[0m, in \u001b[0;36mget\u001b[1;34m(url, params, **kwargs)\u001b[0m\n\u001b[0;32m     62\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mget\u001b[39m(url, params\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs):\n\u001b[0;32m     63\u001b[0m     \u001b[39mr\u001b[39m\u001b[39m\"\"\"Sends a GET request.\u001b[39;00m\n\u001b[0;32m     64\u001b[0m \n\u001b[0;32m     65\u001b[0m \u001b[39m    :param url: URL for the new :class:`Request` object.\u001b[39;00m\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m     70\u001b[0m \u001b[39m    :rtype: requests.Response\u001b[39;00m\n\u001b[0;32m     71\u001b[0m \u001b[39m    \"\"\"\u001b[39;00m\n\u001b[1;32m---> 73\u001b[0m     \u001b[39mreturn\u001b[39;00m request(\u001b[39m\"\u001b[39m\u001b[39mget\u001b[39m\u001b[39m\"\u001b[39m, url, params\u001b[39m=\u001b[39mparams, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs)\n",
      "File \u001b[1;32mc:\\Users\\graha\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\requests\\api.py:59\u001b[0m, in \u001b[0;36mrequest\u001b[1;34m(method, url, **kwargs)\u001b[0m\n\u001b[0;32m     55\u001b[0m \u001b[39m# By using the 'with' statement we are sure the session is closed, thus we\u001b[39;00m\n\u001b[0;32m     56\u001b[0m \u001b[39m# avoid leaving sockets open which can trigger a ResourceWarning in some\u001b[39;00m\n\u001b[0;32m     57\u001b[0m \u001b[39m# cases, and look like a memory leak in others.\u001b[39;00m\n\u001b[0;32m     58\u001b[0m \u001b[39mwith\u001b[39;00m sessions\u001b[39m.\u001b[39mSession() \u001b[39mas\u001b[39;00m session:\n\u001b[1;32m---> 59\u001b[0m     \u001b[39mreturn\u001b[39;00m session\u001b[39m.\u001b[39mrequest(method\u001b[39m=\u001b[39mmethod, url\u001b[39m=\u001b[39murl, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs)\n",
      "File \u001b[1;32mc:\\Users\\graha\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\requests\\sessions.py:587\u001b[0m, in \u001b[0;36mSession.request\u001b[1;34m(self, method, url, params, data, headers, cookies, files, auth, timeout, allow_redirects, proxies, hooks, stream, verify, cert, json)\u001b[0m\n\u001b[0;32m    582\u001b[0m send_kwargs \u001b[39m=\u001b[39m {\n\u001b[0;32m    583\u001b[0m     \u001b[39m\"\u001b[39m\u001b[39mtimeout\u001b[39m\u001b[39m\"\u001b[39m: timeout,\n\u001b[0;32m    584\u001b[0m     \u001b[39m\"\u001b[39m\u001b[39mallow_redirects\u001b[39m\u001b[39m\"\u001b[39m: allow_redirects,\n\u001b[0;32m    585\u001b[0m }\n\u001b[0;32m    586\u001b[0m send_kwargs\u001b[39m.\u001b[39mupdate(settings)\n\u001b[1;32m--> 587\u001b[0m resp \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39msend(prep, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39msend_kwargs)\n\u001b[0;32m    589\u001b[0m \u001b[39mreturn\u001b[39;00m resp\n",
      "File \u001b[1;32mc:\\Users\\graha\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\requests\\sessions.py:745\u001b[0m, in \u001b[0;36mSession.send\u001b[1;34m(self, request, **kwargs)\u001b[0m\n\u001b[0;32m    742\u001b[0m         \u001b[39mpass\u001b[39;00m\n\u001b[0;32m    744\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m stream:\n\u001b[1;32m--> 745\u001b[0m     r\u001b[39m.\u001b[39;49mcontent\n\u001b[0;32m    747\u001b[0m \u001b[39mreturn\u001b[39;00m r\n",
      "File \u001b[1;32mc:\\Users\\graha\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\requests\\models.py:899\u001b[0m, in \u001b[0;36mResponse.content\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    897\u001b[0m         \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_content \u001b[39m=\u001b[39m \u001b[39mNone\u001b[39;00m\n\u001b[0;32m    898\u001b[0m     \u001b[39melse\u001b[39;00m:\n\u001b[1;32m--> 899\u001b[0m         \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_content \u001b[39m=\u001b[39m \u001b[39mb\u001b[39;49m\u001b[39m\"\u001b[39;49m\u001b[39m\"\u001b[39;49m\u001b[39m.\u001b[39;49mjoin(\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49miter_content(CONTENT_CHUNK_SIZE)) \u001b[39mor\u001b[39;00m \u001b[39mb\u001b[39m\u001b[39m\"\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[0;32m    901\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_content_consumed \u001b[39m=\u001b[39m \u001b[39mTrue\u001b[39;00m\n\u001b[0;32m    902\u001b[0m \u001b[39m# don't need to release the connection; that's been handled by urllib3\u001b[39;00m\n\u001b[0;32m    903\u001b[0m \u001b[39m# since we exhausted the data.\u001b[39;00m\n",
      "File \u001b[1;32mc:\\Users\\graha\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\requests\\models.py:818\u001b[0m, in \u001b[0;36mResponse.iter_content.<locals>.generate\u001b[1;34m()\u001b[0m\n\u001b[0;32m    816\u001b[0m     \u001b[39myield from\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mraw\u001b[39m.\u001b[39mstream(chunk_size, decode_content\u001b[39m=\u001b[39m\u001b[39mTrue\u001b[39;00m)\n\u001b[0;32m    817\u001b[0m \u001b[39mexcept\u001b[39;00m ProtocolError \u001b[39mas\u001b[39;00m e:\n\u001b[1;32m--> 818\u001b[0m     \u001b[39mraise\u001b[39;00m ChunkedEncodingError(e)\n\u001b[0;32m    819\u001b[0m \u001b[39mexcept\u001b[39;00m DecodeError \u001b[39mas\u001b[39;00m e:\n\u001b[0;32m    820\u001b[0m     \u001b[39mraise\u001b[39;00m ContentDecodingError(e)\n",
      "\u001b[1;31mChunkedEncodingError\u001b[0m: (\"Connection broken: InvalidChunkLength(got length b'', 0 bytes read)\", InvalidChunkLength(got length b'', 0 bytes read))"
     ]
    }
   ],
   "source": [
    "from importlib import reload \n",
    "reload(uniprot)\n",
    "accs = df.protein_id.unique()\n",
    "print(len(accs))\n",
    "loc_map = {}\n",
    "for i in range(0, len(accs)+1000, 1000):\n",
    "    sent = accs[i:i+1000]\n",
    "    call = uniprot.send_accessions(sent)\n",
    "    resp = uniprot.parse_response(call, wanted_value='gene')\n",
    "    loc_map.update(resp)\n",
    "\n",
    "df.loc[:, 'sub_location'] = df.protein_id.map(loc_map)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.protein_id.drop_duplicates().to_clipboard(index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>From</th>\n",
       "      <th>Entry</th>\n",
       "      <th>Protein names</th>\n",
       "      <th>Gene Names</th>\n",
       "      <th>Length</th>\n",
       "      <th>Subcellular location [CC]</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P10412</td>\n",
       "      <td>P10412</td>\n",
       "      <td>Histone H1.4 (Histone H1b) (Histone H1s-4)</td>\n",
       "      <td>H1-4 H1F4 HIST1H1E</td>\n",
       "      <td>219</td>\n",
       "      <td>SUBCELLULAR LOCATION: Nucleus. Chromosome. Not...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q14978</td>\n",
       "      <td>Q14978</td>\n",
       "      <td>Nucleolar and coiled-body phosphoprotein 1 (14...</td>\n",
       "      <td>NOLC1 KIAA0035 NS5ATP13</td>\n",
       "      <td>699</td>\n",
       "      <td>SUBCELLULAR LOCATION: Nucleus, nucleolus {ECO:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P61927</td>\n",
       "      <td>P61927</td>\n",
       "      <td>60S ribosomal protein L37 (G1.16) (Large ribos...</td>\n",
       "      <td>RPL37</td>\n",
       "      <td>97</td>\n",
       "      <td>SUBCELLULAR LOCATION: Cytoplasm {ECO:0000269|P...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q86VM9</td>\n",
       "      <td>Q86VM9</td>\n",
       "      <td>Zinc finger CCCH domain-containing protein 18 ...</td>\n",
       "      <td>ZC3H18 NHN1</td>\n",
       "      <td>953</td>\n",
       "      <td>SUBCELLULAR LOCATION: Nucleus {ECO:0000250}.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Q9Y6Y8</td>\n",
       "      <td>Q9Y6Y8</td>\n",
       "      <td>SEC23-interacting protein (p125)</td>\n",
       "      <td>SEC23IP MSTP053</td>\n",
       "      <td>1000</td>\n",
       "      <td>SUBCELLULAR LOCATION: Cytoplasmic vesicle, COP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5743</th>\n",
       "      <td>O43505</td>\n",
       "      <td>O43505</td>\n",
       "      <td>Beta-1,4-glucuronyltransferase 1 (EC 2.4.1.-) ...</td>\n",
       "      <td>B4GAT1 B3GNT1 B3GNT6</td>\n",
       "      <td>415</td>\n",
       "      <td>SUBCELLULAR LOCATION: Golgi apparatus membrane...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5744</th>\n",
       "      <td>Q9GZV4</td>\n",
       "      <td>Q9GZV4</td>\n",
       "      <td>Eukaryotic translation initiation factor 5A-2 ...</td>\n",
       "      <td>EIF5A2</td>\n",
       "      <td>153</td>\n",
       "      <td>SUBCELLULAR LOCATION: Cytoplasm {ECO:0000250}....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5745</th>\n",
       "      <td>Q8IY31</td>\n",
       "      <td>Q8IY31</td>\n",
       "      <td>Intraflagellar transport protein 20 homolog (h...</td>\n",
       "      <td>IFT20</td>\n",
       "      <td>132</td>\n",
       "      <td>SUBCELLULAR LOCATION: Golgi apparatus, cis-Gol...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5746</th>\n",
       "      <td>Q9NVQ4</td>\n",
       "      <td>Q9NVQ4</td>\n",
       "      <td>Fas apoptotic inhibitory molecule 1</td>\n",
       "      <td>FAIM FAIM1</td>\n",
       "      <td>179</td>\n",
       "      <td>SUBCELLULAR LOCATION: Cytoplasm {ECO:0000250}.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5747</th>\n",
       "      <td>Q9UJY5</td>\n",
       "      <td>Q9UJY5</td>\n",
       "      <td>ADP-ribosylation factor-binding protein GGA1 (...</td>\n",
       "      <td>GGA1</td>\n",
       "      <td>639</td>\n",
       "      <td>SUBCELLULAR LOCATION: Golgi apparatus, trans-G...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5748 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        From   Entry                                      Protein names  \\\n",
       "0     P10412  P10412         Histone H1.4 (Histone H1b) (Histone H1s-4)   \n",
       "1     Q14978  Q14978  Nucleolar and coiled-body phosphoprotein 1 (14...   \n",
       "2     P61927  P61927  60S ribosomal protein L37 (G1.16) (Large ribos...   \n",
       "3     Q86VM9  Q86VM9  Zinc finger CCCH domain-containing protein 18 ...   \n",
       "4     Q9Y6Y8  Q9Y6Y8                   SEC23-interacting protein (p125)   \n",
       "...      ...     ...                                                ...   \n",
       "5743  O43505  O43505  Beta-1,4-glucuronyltransferase 1 (EC 2.4.1.-) ...   \n",
       "5744  Q9GZV4  Q9GZV4  Eukaryotic translation initiation factor 5A-2 ...   \n",
       "5745  Q8IY31  Q8IY31  Intraflagellar transport protein 20 homolog (h...   \n",
       "5746  Q9NVQ4  Q9NVQ4                Fas apoptotic inhibitory molecule 1   \n",
       "5747  Q9UJY5  Q9UJY5  ADP-ribosylation factor-binding protein GGA1 (...   \n",
       "\n",
       "                   Gene Names  Length  \\\n",
       "0          H1-4 H1F4 HIST1H1E     219   \n",
       "1     NOLC1 KIAA0035 NS5ATP13     699   \n",
       "2                       RPL37      97   \n",
       "3                 ZC3H18 NHN1     953   \n",
       "4             SEC23IP MSTP053    1000   \n",
       "...                       ...     ...   \n",
       "5743     B4GAT1 B3GNT1 B3GNT6     415   \n",
       "5744                   EIF5A2     153   \n",
       "5745                    IFT20     132   \n",
       "5746               FAIM FAIM1     179   \n",
       "5747                     GGA1     639   \n",
       "\n",
       "                              Subcellular location [CC]  \n",
       "0     SUBCELLULAR LOCATION: Nucleus. Chromosome. Not...  \n",
       "1     SUBCELLULAR LOCATION: Nucleus, nucleolus {ECO:...  \n",
       "2     SUBCELLULAR LOCATION: Cytoplasm {ECO:0000269|P...  \n",
       "3          SUBCELLULAR LOCATION: Nucleus {ECO:0000250}.  \n",
       "4     SUBCELLULAR LOCATION: Cytoplasmic vesicle, COP...  \n",
       "...                                                 ...  \n",
       "5743  SUBCELLULAR LOCATION: Golgi apparatus membrane...  \n",
       "5744  SUBCELLULAR LOCATION: Cytoplasm {ECO:0000250}....  \n",
       "5745  SUBCELLULAR LOCATION: Golgi apparatus, cis-Gol...  \n",
       "5746     SUBCELLULAR LOCATION: Cytoplasm {ECO:0000250}.  \n",
       "5747  SUBCELLULAR LOCATION: Golgi apparatus, trans-G...  \n",
       "\n",
       "[5748 rows x 6 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapper = r\"C:\\Users\\graha\\Downloads\\uniprot-download_true_fields_accession_2Cprotein_name_2Cgene_names_2-2022.11.22-00.58.09.52.tsv\"\n",
    "loc_map = pd.read_csv(mapper, delimiter='\\t')\n",
    "loc_map"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.10.4 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "251d0f0bb5f1e19110d9303eadcc132be9bfe80444a6f97cdc41b48d9c9b21c5"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
